### PR TITLE
SignalManager: improve type safety and reduce boilerplate

### PR DESCRIPTION
**Pull request by williamsyang-work**
> ⚠️ **IMPORTANT: Dependency Warning** ⚠️
> 
> This PR introduces type changes that will affect the `vscode-trace-extension` repository.
> **DO NOT MERGE** until maintainers have reviewed and approved the proposed changes.
> A corresponding PR will need to be created in `vscode-trace-extension` to handle these type updates.
>
> 👉 Maintainers: Please review and comment on the proposed type system before work begins on the extension repository.

# Improve SignalManager Type Safety and Developer Experience

## Overview
This PR refactors the SignalManager to provide better TypeScript support and improved developer experience. The changes focus on making the event system more type-safe while reducing boilerplate code.

### 💡 Better Type Inference
```typescript
// Before - No type inference on payload
signalManager.on('THEME_CHANGED', (theme) => {
    console.log(theme.toUpperCase()); // TS doesn't know theme is a string
});

// After - Full type inference
signalManager.on('THEME_CHANGED', (theme) => {
    console.log(theme.toUpperCase()); // TS knows theme is a string
});
```

### 🛡️ Compile-Time Type Safety
```typescript
// Before - TypeScript wouldn't catch this error
signalManager.emit('THEME_CHANGED', 123); // Should be string

// After - TypeScript catches invalid payload types
signalManager.emit('THEME_CHANGED', 123); // Type error!
```

### 📝 Better IDE Support
- Autocomplete for event names
- Parameter type hints for event handlers
- Inline documentation for all methods

### 🧹 Reduced Boilerplate
```typescript
// Before - Required separate fire* method for each event
class SignalManager {
    fireThemeChangedSignal(theme: string): void {
        this.emit(Signals.THEME_CHANGED, theme);
    }
    fireTraceOpenedSignal(trace: Trace): void {
        this.emit(Signals.TRACE_OPENED, trace);
    }
    // ... many more fire* methods
}

// After - Direct emit with type safety
signalManager.emit('THEME_CHANGED', theme);
signalManager.emit('TRACE_OPENED', trace);
```

## Implementation Details
- Consolidated signal types into a single `Signals` interface
- Added generic type constraints for event handlers
- Enhanced JSDoc documentation for core methods
- Maintained backward compatibility with existing event system

### CODE REVIEW

**Review by bhufmann — CHANGES_REQUESTED** [review]
Thanks for the proposal. It looks like a simplification. I had a first glance over the changes and have a general comment using Strings. Please have look.

**Comment by bhufmann** [comment]
FYI @ngondalia

**Comment by bhufmann** [comment]
@williamsyang-work I want to understand the impact to the vscode-trace-extension. I'm currently struggling to make it compile for the external-api (https://github.com/eclipse-cdt-cloud/vscode-trace-extension/blob/e1bfcee1493621b060889a0e79c993c0748931d5/vscode-trace-extension/src/external-api/external-api.ts#L78). 

Any suggestion?

**Comment by bhufmann** [comment]
> @williamsyang-work I want to understand the impact to the vscode-trace-extension. I'm currently struggling to make it compile for the external-api (https://github.com/eclipse-cdt-cloud/vscode-trace-extension/blob/e1bfcee1493621b060889a0e79c993c0748931d5/vscode-trace-extension/src/external-api/external-api.ts#L78).
> 
> I get the following error
> 
> ```
> Argument of type 'string | symbol' is not assignable to parameter of type 'keyof Signals'.
>   Type 'string' is not assignable to type 'keyof Signals'.ts(2345)
> ```
> 
> Any suggestions?

I think we could change the API

```Javascript
 offSignalManagerSignal(event: string | symbol, listener: (...args: unknown[]) => void): void {
        signalManager().off(event, listener);
    },
```

to the following 

```Javascript
offSignalManagerSignal(event: keyof Signals, listener: (...args: unknown[]) => void): void {
        signalManager().off(event, listener);
    },
```

This will break downstream projects. @ngondalia please let us know if that is ok with you.

**Comment by ngondalia** [comment]
> > @williamsyang-work I want to understand the impact to the vscode-trace-extension. I'm currently struggling to make it compile for the external-api (https://github.com/eclipse-cdt-cloud/vscode-trace-extension/blob/e1bfcee1493621b060889a0e79c993c0748931d5/vscode-trace-extension/src/external-api/external-api.ts#L78).
> > I get the following error
> > ```
> > Argument of type 'string | symbol' is not assignable to parameter of type 'keyof Signals'.
> >   Type 'string' is not assignable to type 'keyof Signals'.ts(2345)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any suggestions?
> 
> I think we could change the API
> 
> ```js
>  offSignalManagerSignal(event: string | symbol, listener: (...args: unknown[]) => void): void {
>         signalManager().off(event, listener);
>     },
> ```
> 
> to the following
> 
> ```js
> offSignalManagerSignal(event: keyof Signals, listener: (...args: unknown[]) => void): void {
>         signalManager().off(event, listener);
>     },
> ```
> 
> This will break downstream projects. @ngondalia please let us know if that is ok with you.

This will require a bit of refactoring on our end, but i think it is a good improvement. I am ok with this. :)

**Comment by williamsyang-work** [comment]
Here is a draft PR I made with changes to the `vscode-trace-extension`.  @ngondalia I changed the `ExternalAPI` typing so that the type enforcement will work in downstream projects.

**Review by bhufmann — DISMISSED** [review]
Code looks good to me. I ran it in Theia and with the vscode extension update (https://github.com/eclipse-cdt-cloud/vscode-trace-extension/pull/277) and it works. 

BTW, the license error is taken care of and we can still proceed with merging this PR:
https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/16973

**Comment by bhufmann** [comment]
@ngondalia would you be able to do another review for this one? If not, please let me know and then I'll ask another committer.

**Comment by ngondalia** [comment]
> @ngondalia would you be able to do another review for this one? If not, please let me know and then I'll ask another committer.

Yeah, i can review it - Give me until the end of the day (I would like to try it out with both updates)

**Comment by ngondalia** [comment]
;

> > @ngondalia would you be able to do another review for this one? If not, please let me know and then I'll ask another committer.
> 
> Yeah, i can review it - Give me until the end of the day (I would like to try it out with both updates)

There are some merge conflicts - otherwise LGTM

**Review by bhufmann — CHANGES_REQUESTED** [review]
@williamsyang-work there are linter errors. Please run `yarn format:write` to fix them and then re-submit the changes.

**Review by PatrickTasse — DISMISSED** [review]
I too was worried about not having an enum or constants for the signal names, but the auto-complete does give the list of possible signals.

File: packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx
Comment (williamsyang-work) [reply packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx]: Done.  Sorry for the delay.  @bhufmann

Comment (bhufmann) [reply packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx]: No problem. I'll re-review.

**Review by bhufmann — APPROVED** [review]
It looks good to me. Thanks for the contribution.

**Review by MatthewKhouzam — APPROVED** [review]
I like this approach. I am approving but make sure to merge only when ready.

**Pull request merged**
